src/Strategy/Python/Poetry/PoetryLock.hs

```diff
@@ -0,0 +1,182 @@
+poetryLockPackageSourceCodec =
+  PoetryLockPackageSource
+    <$> Toml.text "type" .= poetryLockPackageSourceType
+    <*> Toml.text "url" .= poetryLockPackageSourceUrl
+    <*> Toml.dioptional (Toml.text "reference") .= poetryLockPackageSourceReference
+    <*> Toml.dioptional (Toml.text "resolved_reference") .= poetryLockPackageSourceResolvedReference
+
+data PoetryLockDependencySpec
+  = TextVersion Text
+  | ObjectVersionSpec ObjectVersion
+  | MultipleObjectVersionSpec [ObjectVersion]
+  deriving (Eq, Ord, Show)
+
+newtype ObjectVersion = ObjectVersion
+  { unObjectVersion :: Text
+  }
+  deriving (Eq, Ord, Show)
+
+objectVersionCodec :: TomlCodec ObjectVersion
+objectVersionCodec =
+  ObjectVersion
+    <$> Toml.text "version" .= unObjectVersion
+
+matchTextVersion :: PoetryLockDependencySpec -> Maybe Text
+matchTextVersion (TextVersion version) = Just version
+matchTextVersion _ = Nothing
+
+matchObjectVersionSpec :: PoetryLockDependencySpec -> Maybe ObjectVersion
+matchObjectVersionSpec (ObjectVersionSpec version) = Just version
+matchObjectVersionSpec _ = Nothing
+
+matchMultipleObjectVersionSpec :: PoetryLockDependencySpec -> Maybe [ObjectVersion]
+matchMultipleObjectVersionSpec (MultipleObjectVersionSpec version) = Just version
+matchMultipleObjectVersionSpec _ = Nothing
+
+poetryLockPackagePoetryLockDependencySpecCodec :: Toml.Key -> TomlCodec PoetryLockDependencySpec
+poetryLockPackagePoetryLockDependencySpecCodec key =
+  Toml.dimatch matchTextVersion TextVersion (Toml.text key)
+    <|> Toml.dimatch matchObjectVersionSpec ObjectVersionSpec (Toml.table objectVersionCodec key)
+    <|> Toml.dimatch matchMultipleObjectVersionSpec MultipleObjectVersionSpec (Toml.list objectVersionCodec key)
+
+-- | Maps poetry lock package to map of package name and associated dependency.
+toMap :: [PoetryLockPackage] -> Map PackageName Dependency
+toMap pkgs =
+  M.fromList
+    [ ( PackageName (toLower $ unPackageName $ poetryLockPackageName pkg)
+      , Dependency
+          { dependencyType = toDepType (poetryLockPackageSource pkg)
+          , dependencyName = toDepName pkg
+          , dependencyVersion = toDepVersion pkg
+          , dependencyLocations = toDepLocs pkg
+          , dependencyEnvironments = toDepEnvironment pkg
+          , dependencyTags = M.empty
+          }
+      )
+    | pkg <- pkgs
+    ]
+  where
+    toDepName :: PoetryLockPackage -> Text
+    toDepName plp = case (poetryLockPackageSource plp) of
+      Nothing -> unPackageName $ poetryLockPackageName plp
+      Just plps -> case poetryLockPackageSourceType plps of
+        "legacy" -> unPackageName $ poetryLockPackageName plp
+        _ -> poetryLockPackageSourceUrl plps
+
+    toDepType :: Maybe PoetryLockPackageSource -> DepType
+    toDepType Nothing = PipType
+    toDepType (Just plps) = case poetryLockPackageSourceType plps of
+      "git" -> GitType
+      "url" -> URLType
+      "legacy" -> PipType
+      _ -> UserType
```

Comment:
This happens when user sources local dependencies (e.g local file, local directory). Will add comment clarifying.

Reply:
To clarify, I'm asking about the `_` match - when would we ever create a `UserType` dependency?

Reply:
My understanding of UserType dependency is for, when user specifies a dependency, and consequently no fetcher/user registry is needed. I've used it here to handle all scenarios where we do not have fetcher, but would still like to still annotate dependency in the graph. Esp. when pkg is sourced from local file, or local directory. Maybe my understanding of the UserType is wrong. Is there other DepType I should be using?

Reply:
I don't think that's quite how it works. @scruffystuffs you implemented user-defined dependencies, right? Can you shed some light on this?

I'm pretty sure what we want to do here is log some kind of warning when we see an unrecognized type.

Reply:
This is what happens right now with user type dependency:

**Excerpts from graph:**
```json
          {
            "imports": [],
            "locator": "user+toml-0.10.2.tar.gz$0.10.2"
          }

          {
            "version": {
              "value": "0.10.2",
              "type": "EQUAL"
            },
            "name": "toml-0.10.2.tar.gz",
            "tags": {},
            "locations": [],
            "type": "UserType"
          },
``` 

![image](https://user-images.githubusercontent.com/86321858/126240734-50c1636e-426b-4b36-9f06-8c937484f5aa.png)

Reply:
@scruffystuffs Let me know if there is more graceful to handle this. I'm presuming we still want to show the dependency on the web app.

Reply:
If we can't possibly fetch it, we don't report it.  That may change in the future, but that's probably the correct intuition for now.  

The `user` fetcher is specifically for explicit user-specified dependencies which cannot be fetched, but should be reported, and should only ever come from the `fossa-deps` file.  Technically, that constructor shouldn't exist at all, since we never look at user-custom projects from a discovery/analysis perspective.